File: src/main/java/at/bitfire/dav4android/UrlUtils.java
package at.bitfire.dav4android;

import android.support.annotation.NonNull;
import android.support.annotation.Nullable;

import org.apache.commons.lang3.StringUtils;

import java.net.URI;
import java.net.URISyntaxException;

import okhttp3.HttpUrl;

public class UrlUtils {

    public static boolean equals(@NonNull HttpUrl url1, @NonNull HttpUrl url2) {
        // if okhttp thinks the two URLs are equal, they're in any case
        // (and it's a simple String comparison)
        if (url1.equals(url2))
            return true;

        URI uri1 = url1.uri(), uri2 = url2.uri();
        try {
            URI decoded1 = new URI(uri1.getScheme(), uri1.getSchemeSpecificPart(), uri1.getFragment()),
                decoded2 = new URI(uri2.getScheme(), uri2.getSchemeSpecificPart(), uri2.getFragment());
            return decoded1.equals(decoded2);
        } catch (URISyntaxException e) {
            return false;
        }
    }

    public static String hostToDomain(@Nullable String host) {
        if (host == null)
            return null;

        // remove optional dot at end
        host = StringUtils.removeEnd(host, ".");

        // split into labels
        String labels[] = StringUtils.split(host, '.');
        if (labels.length >= 2) {
            return labels[labels.length - 2] + "." + labels[labels.length - 1];
        } else
            return host;
    }

    public static HttpUrl omitTrailingSlash(@NonNull HttpUrl url) {
        int idxLast = url.pathSize() - 1;
        boolean hasTrailingSlash = "".equals(url.pathSegments().get(idxLast));

        if (hasTrailingSlash)
            return url.newBuilder().removePathSegment(idxLast).build();
        else
            return url;
    }

    public static HttpUrl withTrailingSlash(@NonNull HttpUrl url) {
        int idxLast = url.pathSize() - 1;
        boolean hasTrailingSlash = "".equals(url.pathSegments().get(idxLast));

        if (hasTrailingSlash)
            return url;
        else
            return url.newBuilder().addPathSegment("").build();
    }

}

File: src/main/java/at/bitfire/dav4android/UrlUtils.kt
package at.bitfire.dav4android;

import okhttp3.HttpUrl
import org.apache.commons.lang3.StringUtils
import java.net.URI
import java.net.URISyntaxException

class UrlUtils {
    companion object {

        @JvmStatic
        fun equals(url1: HttpUrl, url2: HttpUrl): Boolean {
            // if okhttp thinks the two URLs are equal, they're in any case
            // (and it's a simple String comparison)
            if (url1 == url2)
                return true

            val uri1 = url1.uri()
            val uri2 = url2.uri()
            try {
                val decoded1 = URI(uri1.scheme, uri1.schemeSpecificPart, uri1.fragment)
                val decoded2 = URI (uri2.scheme, uri2.schemeSpecificPart, uri2.fragment)
                return decoded1 == decoded2
            } catch (e: URISyntaxException) {
                return false
            }
        }

        @JvmStatic
        fun hostToDomain(host: String?): String? {
            if (host == null)
                return null

            // remove optional dot at end
            @Suppress("NAME_SHADOWING")
            val host = StringUtils.removeEnd(host, ".")

            // split into labels
            val labels = StringUtils.split(host, '.')
            return if (labels.size >= 2) {
                labels[labels.size - 2] + "." + labels[labels.size - 1]
            } else
                host
        }

        @JvmStatic
        fun omitTrailingSlash(url: HttpUrl): HttpUrl {
            val idxLast = url.pathSize () - 1
            val hasTrailingSlash = url.pathSegments()[idxLast] == ""

            return if (hasTrailingSlash)
                url.newBuilder().removePathSegment(idxLast).build()
            else
                url
        }

        @JvmStatic
        fun withTrailingSlash(url: HttpUrl): HttpUrl {
            val idxLast = url.pathSize() - 1
            val hasTrailingSlash = url.pathSegments()[idxLast] == ""

            return if (hasTrailingSlash)
                url
            else
                url.newBuilder().addPathSegment("").build()
        }

    }
}
